File: sdk-node/src/providerModelValidation.ts
import { sdkNodeConfig } from "./config.js";

export class RHEONICValidationError extends Error {
  public readonly provider: string;
  public readonly model: string;
  public readonly expectedProviders: readonly string[];

  public constructor(message: string, provider: string, model: string, expectedProviders: readonly string[]) {
    super(message);
    this.name = "RHEONICValidationError";
    this.provider = provider;
    this.model = model;
    this.expectedProviders = expectedProviders;
  }
}

export function validateProviderModel(provider: string, model: string | null | undefined): void {
  const normalizedProvider = provider.trim().toLowerCase();
  if (!normalizedProvider) {
    throw new RHEONICValidationError(
      "RHEONIC: provider must be explicitly provided.",
      provider,
      String(model ?? ""),
      sdkNodeConfig.supportedProviders,
    );
  }

  if (!sdkNodeConfig.supportedProviders.includes(normalizedProvider as (typeof sdkNodeConfig.supportedProviders)[number])) {
    throw new RHEONICValidationError(
      `RHEONIC: unsupported provider: ${provider}`,
      provider,
      String(model ?? ""),
      sdkNodeConfig.supportedProviders,
    );
  }

  const normalizedModel = typeof model === "string" ? model.trim() : "";
  if (!normalizedModel) {
    throw new RHEONICValidationError(
      `RHEONIC: model must be explicitly provided for provider ${normalizedProvider}.`,
      normalizedProvider,
      String(model ?? ""),
      sdkNodeConfig.supportedProviders,
    );
  }
}

File: sdk-python/src/rheonic/provider_model_validation.py
from rheonic.config import sdk_config
from rheonic.protect_engine import RHEONICValidationError


def validate_provider_model(provider: str, model: str | None) -> None:
    normalized_provider = provider.strip().lower()
    if not normalized_provider:
        raise RHEONICValidationError("RHEONIC: provider must be explicitly provided.")

    if normalized_provider not in sdk_config.supported_providers:
        raise RHEONICValidationError(f"RHEONIC: unsupported provider: {provider}")

    normalized_model = (model or "").strip()
    if not normalized_model:
        raise RHEONICValidationError(f"RHEONIC: model must be explicitly provided for provider {normalized_provider}.")
